### DISCUSSION TITLE
Dynamic Content Creation Questions

### DISCUSSION THREAD

**Discussion by xfow** [elkowar/eww#656] [Q&A]
Hi there,

I've just started with EWW and I'm creating a small widget which interacts with an API to dynamically list out data. I have a few questions as I've hit some roadblocks.

**Dynamically Creating Revealers**

So for a basic toggle button revealer, I can use:

```
(defwidget clickbox []
     (button :onclick "eww update revadd=${revadd ? false : true} :width 200 :height 50) 
                  (box :orientation "h"
                              :width 200
			      :height 500
                         (revealer :reveal revadd
			           :transition "slidedown"
				   (box 
				       (input :width 150
				              :height 100
					      :style "background-color: black;"
				    	)
				   )
)
(defvar revadd false)
```

However, if I define a widget, so I can add it to a literal, the toggle doesn't seem to work the way I want.

```
(defwidget clickbox [vtitle vclass revadd]
   (box :orientation "v"
         :space-evenly false
            (box :orientation "h"
		         :height 50
		         :width 200
	             :space-evenly false
		         (box :width 10
		              :height 50
		              :class "${vclass}"
		         )
		    (eventbox :width 190
		              :height 50
		              :class "vclass"
		              :onclick "eww update revadd=${ revadd ? false : true }"
		              (label :text "${vtitle}"
		                     :xalign 0.1
			                 :height 50
			                 :vexpand false
			                 :hexpand false
			                 :tooltip "${vtitle}"
			                 :style "color: black;"
			                 :wrap true
		              )
		
	        )
		
            )
	        (revealer :reveal revadd
		              :transition "slidedown"
		              (box :width 200
		                   :height 20
	                       :style "background-color: black;"
		              )
            )
   
   )
)
```

So the problem I'm having is that the revadd attribute used by onclick is evaluating to false, rather than the variable name. I can't seem to add the variable name I'm looking for in order to get the click working correctly. Something like

`:onclick "eww update ${revaddname}=${ revadd ? false : true }"`

But then defining the variable name is becoming problematic. I need to define it as:

`:revaddname revadd`

And then add the reveal top evaluate the variable name, and then the value it's been set to, something like:

`:reveal ${${revaddname}}`

I can't find the structure to create dynamic variables in this way, instead I can only define the values accepted by predefined variables.

**Listener Not Working As Expected**

The literal I'm using interacts with an API to pull information. The script is written in Python. When I define the variable as a poll, the script works fine. Here's the script (approximately)

```
#! /usr/bin/env python3
import httpx
import json
import time
import os

home = os.environ["HOME"]

with open(f'{home}/.config/eww/v.conf', "r") as w:
    userdetails = json.load(w)

accesskey = userdetails["AccessKey"]

initial = ""
def getv():
    genbox = ""
    req = httpx.get(
            url,
            headers={
                "API-ACCESS-KEY": accesskey,
                "Content-Type": "application/json"
            }
    )

    v = req.json()['v']
    v.sort(key = lambda d: float(d["v"]), reverse = True)
    genbox = '(box :orientation "v" '
    for i in v:
        genbox += f'(vbox :vtitle "{i["v_title"]}" '
        if float(i["vscore"]) >= 9:
            genbox += ':vclass "critical" '
        elif float(i["vscore"]) >= 7:
            genbox += ':vclass "high" '
        elif float(i["vscore"]) >= 4:
            genbox += ':vclass "medium" '
        elif float(i["vscore"]) >= 0.1:
            genbox += ':vclass "low" '
        else:
            genbox += ':vclass "info" '
        genbox += ':revaddname "test" :revadd false)'
    genbox += ')'
    return genbox

print(genbox())
```

However, I want a listener instead, for the reason that if I'm adding in revealers, expanders etc, then I want the values for them initially set to false, and I don't want the poll to reload the widgets and set the values to false every time it polls the server. Instead, I want to use a listener, so I've modified the script. I'll be adding in conditionals later to check whether there's been any changes on the API. However, for now, I've just added an infinite loop to to the Python script, which continuosly prints, so in theory it should act exactly the same as a poll.

```
#! /usr/bin/env python3
import httpx
import json
import time
import os

home = os.environ["HOME"]

with open(f'{home}/.config/eww/v.conf', "r") as w:
    userdetails = json.load(w)

accesskey = userdetails["AccessKey"]

initial = ""
def getv():
    genbox = ""
    req = httpx.get(
            url,
            headers={
                "API-ACCESS-KEY": accesskey,
                "Content-Type": "application/json"
            }
    )

    v = req.json()['v']
    v.sort(key = lambda d: float(d["v"]), reverse = True)
    genbox = '(box :orientation "v" '
    for i in v:
        genbox += f'(vbox :vtitle "{i["v_title"]}" '
        if float(i["vscore"]) >= 9:
            genbox += ':vclass "critical" '
        elif float(i["vscore"]) >= 7:
            genbox += ':vclass "high" '
        elif float(i["vscore"]) >= 4:
            genbox += ':vclass "medium" '
        elif float(i["vscore"]) >= 0.1:
            genbox += ':vclass "low" '
        else:
            genbox += ':vclass "info" '
        genbox += ':revaddname "test" :revadd false)'
    genbox += ')'
    return genbox

while True:
	print(genbox())
```

The generated widgets, although completely the same as the ones made when polling, just don't render at all when used in this way.

Is there a reason why and a workaround? I've tried using a bash script to call the Python instead, and I've tried using an os.system to echo the value of genbox instead of using python's print.

**Comment by musjj** [comment]
Refer to: https://github.com/elkowar/eww/discussions/514
